### ISSUE TITLE
Unexpected ERR_UseDefViolation error

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#9154]
```
    void Test5()
    {
        Dummy((System.Func<object, object, bool>) ((o1, o2) => o1 is int x5 && 
                                                               o2 is int x5 && 
                                                               x5 > 0));
    }
```

Observed:

```
    // (33,74): error CS0128: A local variable named 'x5' is already defined in this scope
    //                                                                o2 is int x5 && 
    Diagnostic(ErrorCode.ERR_LocalDuplicate, "x5").WithArguments("x5").WithLocation(33, 74),
    // (34,64): error CS0165: Use of unassigned local variable 'x5'
    //                                                                x5 > 0));
    Diagnostic(ErrorCode.ERR_UseDefViolation, "x5").WithArguments("x5").WithLocation(34, 64)
```

**Comment by gafter** [comment]
Also

``` c#
        switch (true)
        {
            case new object() is int x1:
                System.Console.WriteLine(x1); // improper def assignment error
                break;
        }
```

**Comment by gafter** [comment]
Fixed in https://github.com/dotnet/roslyn/issues/25774
